File: src/main/scala/fortress/transformers/ClosureEliminationTransformer.java
package fortress.transformers;

import fortress.tfol.*;
import java.util.Set;
import java.util.HashSet;
import fortress.data.NameGenerator;
import fortress.data.SubIntNameGenerator;
import fortress.tfol.operations.ClosureEliminator;

/** Replaces transitive closure terms with a term representing the application of a new function
 but with same arguments. **/
public class ClosureEliminationTransformer implements TheoryTransformer {
    @Override
    public Theory apply(Theory theory) {
        Set<String> forbiddenNames = new HashSet<>();
        
        for(Type type : theory.getTypes()) {
            forbiddenNames.add(type.getName());
        }
        
        for(FuncDecl fdecl : theory.getFunctionDeclarations()) {
            forbiddenNames.add(fdecl.getName());
        }
        
        for(AnnotatedVar c : theory.getConstants()) {
            forbiddenNames.add(c.getName());
        }
        
        for(Term axiom : theory.getAxioms()) {
            forbiddenNames.addAll(axiom.allSymbolsJava());
        }
        
        NameGenerator nameGenerator = new SubIntNameGenerator(forbiddenNames, 0);
        
        Theory result = theory.withoutAxioms();
        for(Term axiom : theory.getAxioms()) {
            ClosureEliminator closureEliminator = new ClosureEliminator(axiom, result.getSignature(), nameGenerator);
            Term newAxiom = closureEliminator.convert();
            result = result.withFunctionDeclarations(closureEliminator.getClosureFunctions());
            result = result.withAxiom(newAxiom);
        }
        
        return result;
    }
    
    @Override
    public String getName() {
        return "Closure Elimination Transformer";
    }
}

File: src/main/scala/fortress/transformers/ClosureEliminationTransformer.scala
package fortress.transformers

import fortress.msfol._
import fortress.data.NameGenerator
import fortress.data.IntSuffixNameGenerator
import fortress.operations.ClosureEliminator
import fortress.operations.TheoryOps._
import fortress.interpretation.Interpretation

/** Replaces transitive closure terms with a term representing the application of a new relation
 but with same arguments. **/
class ClosureEliminationTransformer extends ProblemStateTransformer {
    override def apply(problemState: ProblemState): ProblemState = problemState match {
        case ProblemState(theory, scopes, skc, skf, rangeRestricts, unapplyInterp) => {
            val forbiddenNames = scala.collection.mutable.Set[String]()
            
            for(sort <- theory.sorts) {
                forbiddenNames += sort.name
            }
            
            for(fdecl <- theory.functionDeclarations) {
                forbiddenNames += fdecl.name
            }
            
            for(constant <- theory.constants) {
                forbiddenNames += constant.name
            }
            
            val nameGenerator = new IntSuffixNameGenerator(forbiddenNames.toSet, 0)
            
            var resultTheory = theory.withoutAxioms
            for(axiom <- theory.axioms) {
                val closureEliminator = new ClosureEliminator(axiom, resultTheory.signature, scopes, nameGenerator)
                val newAxiom = closureEliminator.convert()
                resultTheory = resultTheory.withFunctionDeclarations(closureEliminator.getClosureFunctions.toList)
                resultTheory = resultTheory.withAxioms(closureEliminator.getClosureAxioms.toList)
                resultTheory = resultTheory.withAxiom(newAxiom)
            }
            
            ProblemState(resultTheory, scopes, skc, skf, rangeRestricts, unapplyInterp)
        }
    }
    
    override def name: String = "Closure Elimination Transformer"
    
}